monai/data/dataset.py

```diff
@@ -1348,3 +1348,37 @@ def __init__(
             dfs=dfs, row_indices=row_indices, col_names=col_names, col_types=col_types, col_groups=col_groups, **kwargs
         )
         super().__init__(data=data, transform=transform)
+
+
+class HashCacheDataset(CacheDataset):
+    """
+    Extend from `CacheDataset` to support only caching unique items in the datset.
+    It computes hash value of input data as the key to save cache, if key exists, avoid saving duplicated content.
+    Can help save memory when the dataset has duplicated items or augmented dataset.
+    The `cache_num` or `cache_rate` are computed against only on the unique items of the dataset.
+
+    Args:
+        data: input data to load and transform to generate dataset for model.
+        hash_func: a callable to compute hash from data items to be cached.
+            defaults to `monai.data.utils.pickle_hashing`.
+        kwargs: other arguments of `CacheDataset` except for `data`.
+
+    """
+
+    def __init__(self, data: Sequence, hash_func: Callable[..., bytes] = pickle_hashing, **kwargs) -> None:
+        self.hash_func = hash_func
+        mapping = {self.hash_func(v): v for v in data}
+        # only compute cache for the unique items of dataset
+        super().__init__(data=list(mapping.values()), **kwargs)
+        self._cache_keys = list(mapping)[: self.cache_num]
```

Comment:
Thanks, this looks straightforward, why not merge this as an option to the cache dataset and by default enable it? With this, perhaps the original cachedatset is less useful.

Reply:
Hi @wyli ,

The reason I didn't add it in the CacheDataset:
1. It may be faster to use indices directly(no hash computation and no need to search the key name in the list)? Performance is very important for CacheDataset.
2. I am not sure how it affect the subclasses, like `SmartCacheDataset` replacement logic with duplicated items..

Anyway, if you think it's better to update the `CacheDataset` directly, I can think about it more.
What do you think?

Thanks.

Reply:
Sure if the solution is simple and efficient I think we should merge and make it default. Otherwise the current is ok

Reply:
Hi @wyli ,

Thanks for your review.
I tried to leverage this Hash logic for the subclass `SmartCacheDataset`, but it's really not clear to develop the "replacement" behavior, for example: if we still use all the data for training, everytime we do replacement, the `unique cache` size may be different, so the training may out of memory at some epoch. And in this case, the definition of `cache_num` will be different from `CacheDataset` which is only the unique cache size.
I would suggest current independent `HashCacheDataset` for this use case, if we get more feature requests for it, we can consider to revisit it and discuss how to make it work for all the subclasses of `CacheDataset`.
What do you think?

Thanks in advance.

Reply:
I agree we can consider the samrtcachedataset later. I think the idea here is to have an option in the new cachedataset to switch between the original cachedataset vs hash-based cachedataset. it should be non-breaking and doesn't affect the subclass of cachedataset. or do I misunderstand?

Reply:
Hi @wyli ,

Thanks for your suggestion.
And I also found a bug in the previous `CacheDataset`, when calling `set_data()` to update data and cache, the `self.cache_num` is not updated...
Let me try to integrate HashCacheDataset into CacheDataset directly.

Thanks.

Reply:
Updated in the latest commit.

Thanks.